kornia/geometry/boxes.py

```diff
@@ -0,0 +1,705 @@
+from typing import Optional, Tuple
+
+import torch
+
+from kornia.geometry.linalg import transform_points
+
+__all__ = ["Boxes", "Boxes3D"]
+
+
+def _is_floating_point_dtype(dtype: torch.dtype) -> bool:
+    return dtype in (torch.float16, torch.float32, torch.float64, torch.bfloat16, torch.half)
+
+
+def _transform_boxes(boxes: torch.Tensor, M: torch.Tensor) -> torch.Tensor:
+    """Transforms 3D and 2D in kornia format by applying the transformation matrix M. Boxes and the transformation
+    matrix could be batched or not.
+
+    Args:
+        boxes: 2D quadrilaterals or 3D hexahedrons in kornia format.
+        M: the transformation matrix of shape :math:`(3, 3)` or :math:`(B, 3, 3)` for 2D and :math:`(4, 4)` or
+            :math:`(B, 4, 4)` for 3D hexahedron.
+    """
+    M = M if M.is_floating_point() else M.float()
+
+    # Work with batch as kornia.transform_points only supports a batch of points.
+    boxes_per_batch, n_points_per_box, coordinates_dimension = boxes.shape[-3:]
+    points = boxes.view(-1, n_points_per_box * boxes_per_batch, coordinates_dimension)
+    M = M if M.ndim == 3 else M.unsqueeze(0)
+
+    if points.shape[0] != M.shape[0]:
+        raise ValueError(
+            f"Batch size mismatch. Got {points.shape[0]} for boxes and {M.shape[0]} for the transformation matrix."
+        )
+
+    transformed_boxes: torch.Tensor = transform_points(M, points)
+    transformed_boxes = transformed_boxes.view_as(boxes)
+    return transformed_boxes
+
+
+def _boxes_to_polygons(
+    xmin: torch.Tensor, ymin: torch.Tensor, width: torch.Tensor, height: torch.Tensor
+) -> torch.Tensor:
+    if not xmin.ndim == ymin.ndim == width.ndim == height.ndim == 2:
+        raise ValueError("We expect to create a batch of 2D boxes (quadrilaterals) in vertices format (B, N, 4, 2)")
+
+    # Create (B,N,4,2) with all points in top left position of boxes
+    polygons = torch.zeros((xmin.shape[0], xmin.shape[1], 4, 2), device=xmin.device, dtype=xmin.dtype)
+    polygons[..., 0] = xmin.unsqueeze(-1)
+    polygons[..., 1] = ymin.unsqueeze(-1)
+    # Shift top-right, bottom-right, bottom-left points to the right coordinates
+    polygons[..., 1, 0] += width - 1  # Top right
+    polygons[..., 2, 0] += width - 1  # Bottom right
+    polygons[..., 2, 1] += height - 1  # Bottom right
+    polygons[..., 3, 1] += height - 1  # Bottom left
```

Comment:
@hal-314 @edgarriba @shijianjian 

Why is this -1 required?

```
from kornia.geometry.boxes import _boxes_to_polygons

xmin = torch.tensor([[1,2]])
ymin = torch.tensor([[2,4]])
width = torch.tensor([[2,2]])
height = torch.tensor([[4,4]])
res = _boxes_to_polygons(xmin, ymin, width, height)
```
With the -1, this returns polygons as 

```
tensor([[[[1., 2.],
          [2., 2.],
          [2., 5.],
          [1., 5.]],

         [[2., 4.],
          [3., 4.],
          [3., 7.],
          [2., 7.]]]])
```
Which aren't the quadrilateral points. Was this intentional?

Reply:
Removing the -1 would get you the correct quadrilateral points

```
tensor([[[[1., 2.],
          [3., 2.],
          [3., 6.],
          [1., 6.]],

         [[2., 4.],
          [4., 4.],
          [4., 8.],
          [2., 8.]]]])
```